viewing_party/party.py

```diff
@@ -1,23 +1,152 @@
+    # print("AAAAAAAAAAA")
+
+    user_data["watched"].append(movie)
+
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data, movie):
+    for movies in user_data["watchlist"]:
+        if movies["title"] == movie:
+            user_data["watched"].append(movies)
+            user_data["watchlist"].remove(movies)
+
+    return user_data
+
+
+user_data = {"watched": [], "watchlist": [{
+    "title": "MOVIE_TITLE_1",
+    "genre": "GENRE_1",
+    "rating": "RATING_1"
+}]}
+movie = {
+    "title": "MOVIE_TITLE_1",
+    "genre": "GENRE_1",
+    "rating": "RATING_1"
+}
+print(watch_movie(user_data, movie))
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    total = 0
+    count = 0
+    if not user_data["watched"]:
+        return 0
+    for data in user_data["watched"]:
+        total += data["rating"]
+        count += 1
+    average = total / count
+
+    return average
+
+
+def get_most_watched_genre(user_data):
+    genres = []
+    if not user_data["watched"]:
+        return None
+    for data in user_data["watched"]:
+        genres.append(data["genre"])
+    top_genre = ""
+    top_count = 0
+    for genre in genres:
+        genre_count = genres.count(genre)
+        if (genre_count > top_count):
+            top_count = genre_count
+            top_genre = genre
+    return top_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    if not user_data["watched"]:
+        return []
+    friends_watched = []
+    user_unique_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched.append(movie["title"])
+    for data in user_data["watched"]:
+        if data["title"] not in friends_watched:
+            user_unique_movies.append(data)
+    return user_unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    user_watched = user_data["watched"]
+    friends_unique_movie = []
+    for friend in user_data["friends"]:
+        for data in friend["watched"]:
+            if data not in user_watched and data not in friends_unique_movie:
+                friends_unique_movie.append(data)
+    return friends_unique_movie
+
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
+
 # -----------------------------------------
+def get_available_recs(user_data):
+    # subscriptions - what user has
+    # host - where the movie is hosted
+    #  list of movies that friends have watch and user has not
+    # movies must be in host the user has
+    recommendations = []
+    print(user_data["subscriptions"])
+    friends_unique = get_friends_unique_watched(user_data)

```

Comment:
👍 The description for this function states that the results should only include movies that a friend has watched, but the user has not. This sounds a lot like what `get_friends_unique_watched` calculates!

By starting with a call to `get_friends_unique_watched` to get only the movies that the friends have watched, but the user has not, the only thing we have left to do is to filter those movies down further to the ones who share a host with one of the user subscriptions!

This makes our code much more DRY!